### PR TITLE
DeviceTwin: Adding test cases for communicate.go

### PR DESCRIPTION
**Pull request by lidiyag**
This adds unit test cases for edge/pkg/devicetwin/dtmanager/communicate.go

**What type of PR is this?**
 /kind test

**What this PR does / why we need it**:
This PR adds test cases for communicate.go

### CODE REVIEW

File: edge/pkg/devicetwin/dtmanager/communicate_test.go [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+			if !reflect.DeepEqual(err, tt.wantErr) {
+				t.Errorf("CommWorker.checkConfirm() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("CommWorker.checkConfirm() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
```
Comment (sids-b) [inline edge/pkg/devicetwin/dtmanager/communicate_test.go]: Add a blank line at the end

File: edge/pkg/devicetwin/dtmanager/communicate_test.go [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+				Worker: tt.Worker,
+				Group:  tt.Group,
+			}
+			go cw.Start()
+		})
+	}
+}
+
+// Test_dealSendToEdge is function to test dealsendToedge().
+func Test_dealSendToEdge(t *testing.T) {
```
Comment (sids-b) [inline edge/pkg/devicetwin/dtmanager/communicate_test.go]: Camel case please.
Applicable to all cases below

File: edge/pkg/devicetwin/dtmanager/communicate_test.go (hunk 1 above)
Comment (lidiyag) [reply edge/pkg/devicetwin/dtmanager/communicate_test.go]: Updated

File: edge/pkg/devicetwin/dtmanager/communicate_test.go (hunk 2 above)
Comment (lidiyag) [reply edge/pkg/devicetwin/dtmanager/communicate_test.go]: Updated

File: edge/pkg/devicetwin/dtmanager/communicate_test.go [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+			context:  dtContexts,
+			resource: "",
+			msg:      &model.Message{},
+			wantErr:  nil,
+			want:     nil,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got, err := dealSendToEdge(test.context, test.resource, test.msg)
```
Comment (rohitsardesai83) [inline edge/pkg/devicetwin/dtmanager/communicate_test.go]: the dealSendToEdge method always returns nil, nil. Are return values required here ? Please check and if not needed , adjust your code and test case accordingly

File: edge/pkg/devicetwin/dtmanager/communicate_test.go [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+		{
+			name:     "dealSendToCloudTest-StateConnected",
+			context:  dtContextStateConnected,
+			resource: "",
+			msg:      "",
+			want:     nil,
+			wantErr:  errors.New("msg not Message type"),
+		},
+		{
+			name:     "dealSendToCloudTest-ActualMsg",
```
Comment (rohitsardesai83) [inline edge/pkg/devicetwin/dtmanager/communicate_test.go]: In the case of a correct message being sent , the context's confirmMap stores the msg id along with the DTMessage...how can we test this ?

File: edge/pkg/devicetwin/dtmanager/communicate_test.go [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+			Group: "",
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			cw := CommWorker{
+				Worker: test.Worker,
+				Group:  test.Group,
+			}
+			go cw.Start()
```
Comment (rohitsardesai83) [inline edge/pkg/devicetwin/dtmanager/communicate_test.go]: the start method has 3 select cases, one where receiver chan gets a msg, second where timeout occurs, third where the hearbeat happens. can we have separate test cases to test 2 of these , maybe the timeout case we can skip.

Comment (rohitsardesai83) [reply edge/pkg/devicetwin/dtmanager/communicate_test.go]: the heartbeat chan is passed 2 messages , one is "ping" to keep on hearbeat alive , second is "stop" to stop the heartbeat. we need to have tests for these.

File: edge/pkg/devicetwin/dtmanager/communicate_test.go [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+			name:    "dealConfirmTest-WrongMsg",
+			context: dtContext,
+			msg:     "",
+			want:    nil,
+			wantErr: errors.New("CommModule deal confirm, type not correct"),
+		},
+		{
+			name:    "dealConfirmTest-CorrectMsg",
+			context: dtContext,
+			msg:     &model.Message{},
```
Comment (rohitsardesai83) [inline edge/pkg/devicetwin/dtmanager/communicate_test.go]: the function of dealConfirm is to check the message type and then delete the parentMsgID from the ConfirmMap. 
we need a way to validate whether the parentMsgID is deleted correctly. please add a test case for this.

File: edge/pkg/devicetwin/dtmanager/communicate_test.go [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+			}
+			if !reflect.DeepEqual(got, test.want) {
+				t.Errorf("dealConfirm() = %v, want %v", got, test.want)
+			}
+		})
+	}
+}
+
+// TestDetailRequest is function to test detailRequest().
+func TestDetailRequest(t *testing.T) {
```
Comment (rohitsardesai83) [inline edge/pkg/devicetwin/dtmanager/communicate_test.go]: why do we have a separate test case for testing the detailRequest method ? This is called as a part of dealLifecycle method. the test for dealLifecycle should have a case where this method is called.

File: edge/pkg/devicetwin/dtmanager/communicate_test.go [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+		})
+	}
+}
+
+// TestCheckConfirm is function to test checkConfirm().
+func TestCheckConfirm(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContext, _ := dtcontext.InitDTContext(mainContext)
+	dtContext.ConfirmMap.Store("key", &dttype.DTMessage{})
+	dtContext.ConfirmMap.Store("key", &dttype.DTMessage{Action: dtcommon.SendToCloud})
```
Comment (rohitsardesai83) [inline edge/pkg/devicetwin/dtmanager/communicate_test.go]: we overwrite the same key ? how will it help test the behaviour of the confirm method ?

**Comment by sids-b** [comment]
@lidiyag Please resolve the comments so that we can go-ahead with merging.

File: edge/pkg/devicetwin/dtmanager/communicate_test.go (hunk 3 above)
Comment (lidiyag) [reply edge/pkg/devicetwin/dtmanager/communicate_test.go]: Updated

File: edge/pkg/devicetwin/dtmanager/communicate_test.go (hunk 4 above)
Comment (lidiyag) [reply edge/pkg/devicetwin/dtmanager/communicate_test.go]: Updated

File: edge/pkg/devicetwin/dtmanager/communicate_test.go (hunk 5 above)
Comment (lidiyag) [reply edge/pkg/devicetwin/dtmanager/communicate_test.go]: Updated

File: edge/pkg/devicetwin/dtmanager/communicate_test.go (hunk 6 above)
Comment (lidiyag) [reply edge/pkg/devicetwin/dtmanager/communicate_test.go]: Updated

File: edge/pkg/devicetwin/dtmanager/communicate_test.go (hunk 8 above)
Comment (lidiyag) [reply edge/pkg/devicetwin/dtmanager/communicate_test.go]: Updated

File: edge/pkg/devicetwin/dtmanager/communicate_test.go [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+limitations under the License.
+*/
+
+package dtmanager
+
+import (
+	"errors"
+	"github.com/kubeedge/kubeedge/common/beehive/pkg/core/context"
+	"github.com/kubeedge/kubeedge/common/beehive/pkg/core/model"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcommon"
```
Comment (sids-b) [inline edge/pkg/devicetwin/dtmanager/communicate_test.go]: seperate golang, kubeedge and vendor packages.

File: edge/pkg/devicetwin/dtmanager/communicate_test.go [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+		Worker Worker
+		Group  string
+	}{
+		{
+			name: "StartTest-ActionPresentInActionCallback",
+			Worker: Worker{
+				ReceiverChan: receiveChanActionPresent,
+				DTContexts:   dtContextStateConnected,
+			},
+			Group: "",
```
Comment (sids-b) [inline edge/pkg/devicetwin/dtmanager/communicate_test.go]: default value of a string is ""....
applicable to all similar cases below.

File: edge/pkg/devicetwin/dtmanager/communicate_test.go [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+			}
+		})
+	}
+}
+
+// TestStartHeartBeat is function to test Start() when value is passed in HeartBeatChan.
+func TestStartHeartBeat(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContexts, _ := dtcontext.InitDTContext(mainContext)
+	heartChanStop := make(chan interface{}, 1)
```
Comment (sids-b) [inline edge/pkg/devicetwin/dtmanager/communicate_test.go]: Do we need a buffered channel here ?

File: edge/pkg/devicetwin/dtmanager/communicate_test.go [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+			Group: "",
+		},
+	}
+	for _, test := range tests {
+		cw := CommWorker{
+			Worker: test.Worker,
+			Group:  test.Group,
+		}
+		go cw.Start()
+		time.Sleep(1 * time.Millisecond)
```
Comment (sids-b) [inline edge/pkg/devicetwin/dtmanager/communicate_test.go]: can we keep the 3 steps above inside t.Run ??
applicable to all similar cases below

File: edge/pkg/devicetwin/dtmanager/communicate_test.go [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			cw := CommWorker{
+				Worker: test.Worker,
+				Group:  test.Group,
+			}
+			cw.checkConfirm(test.context, test.msg)
+			for _, test := range tests {
```
Comment (sids-b) [inline edge/pkg/devicetwin/dtmanager/communicate_test.go]: Do we need to loop twice over the tests array ? we also have a nested t.Run.....

File: edge/pkg/devicetwin/dtmanager/communicate_test.go (hunk 9 above)
Comment (lidiyag) [reply edge/pkg/devicetwin/dtmanager/communicate_test.go]: Updated

File: edge/pkg/devicetwin/dtmanager/communicate_test.go (hunk 11 above)
Comment (lidiyag) [reply edge/pkg/devicetwin/dtmanager/communicate_test.go]: Yes

File: edge/pkg/devicetwin/dtmanager/communicate_test.go (hunk 12 above)
Comment (lidiyag) [reply edge/pkg/devicetwin/dtmanager/communicate_test.go]: Updated

File: edge/pkg/devicetwin/dtmanager/communicate_test.go (hunk 13 above)
Comment (lidiyag) [reply edge/pkg/devicetwin/dtmanager/communicate_test.go]: Updated

File: edge/pkg/devicetwin/dtmanager/communicate_test.go [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+
+package dtmanager
+
+import (
+	"errors"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/kubeedge/kubeedge/common/beehive/pkg/core/context"
```
Comment (sids-b) [inline edge/pkg/devicetwin/dtmanager/communicate_test.go]: this import path doesn't exist anymore. Please change

Comment (lidiyag) [reply edge/pkg/devicetwin/dtmanager/communicate_test.go]: Updated

**Pull request merged**